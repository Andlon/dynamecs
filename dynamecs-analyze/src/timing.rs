use std::collections::hash_map::Entry;
use std::collections::HashMap;
use std::time::{Duration as StdDuration};
use eyre::eyre;
use time::OffsetDateTime;
use crate::{Record, RecordKind};

#[derive(Debug, Clone)]
pub struct SpanTiming {
    span_name: String,
    children: Vec<SpanTiming>,
    self_duration: StdDuration,
}

impl SpanTiming {
    pub fn name(&self) -> &str {
        &self.span_name
    }

    pub fn self_duration(&self) -> StdDuration {
        self.self_duration
    }

    pub fn total_duration(&self) -> StdDuration {
        let mut duration = self.self_duration;
        for child in &self.children {
            duration += child.total_duration();
        }
        duration
    }

    pub fn children(&self) -> &[SpanTiming] {
        &self.children
    }
}

pub fn accumulate_timings(records: impl IntoIterator<Item=Record>) -> eyre::Result<Vec<SpanTiming>> {
    accumulate_timings_(records.into_iter())
}

fn accumulate_timings_(records: impl Iterator<Item=Record>) -> eyre::Result<Vec<SpanTiming>> {
    let mut span_timer = SpanTimer::default();
    for record in records {
        if [RecordKind::SpanEnter, RecordKind::SpanExit].contains(&record.kind()) {
            let mut span_names = record.spans().iter().map(|span| span.name().to_string()).collect();
            match record.kind {
                RecordKind::SpanEnter => span_timer.enter_span(span_names, *record.timestamp())?,
                RecordKind::SpanExit => {
                    // Exit records do not record their own span in their list of spans,
                    // so we must add this to the span list in order to match up with
                    // the enter record
                    span_names.push(record.span().name().to_string());
                    span_timer.exit_span(span_names, *record.timestamp())?
                },
                RecordKind::Event => {}
            }
        }
    }
    Ok(span_timer.collect_accumulated_timings())
}

#[derive(Default, Debug)]
struct SpanTimer {
    accumulated_span_timings: HashMap<Vec<String>, StdDuration>,
    open_span_enter_timestamps: HashMap<Vec<String>, OffsetDateTime>
}

impl SpanTimer {
    fn enter_span(&mut self, spans: Vec<String>, timestamp: OffsetDateTime) -> eyre::Result<()> {
        let entry = self.open_span_enter_timestamps.entry(spans);
        match entry {
            Entry::Occupied(_) => Err(eyre!("Tried to enter span that was already entered")),
            Entry::Vacant(entry) => {
                entry.insert(timestamp); Ok(())
            }
        }
    }

    fn exit_span(&mut self, spans: Vec<String>, timestamp: OffsetDateTime) -> eyre::Result<()> {
        let entry = self.open_span_enter_timestamps.entry(spans);
        match entry {
            Entry::Occupied(entry) => {
                let exit_timestamp = timestamp;
                let (spans, enter_timestamp) = entry.remove_entry();
                let duration: StdDuration = (exit_timestamp - enter_timestamp).unsigned_abs();
                let old_duration = self.accumulated_span_timings.entry(spans)
                    .or_insert_with(|| StdDuration::ZERO);
                *old_duration += duration;
                Ok(())
            },
            Entry::Vacant(_) => {
                Err(eyre!("Tried to exit span that was not already entered"))
            }
        }
    }

    fn collect_accumulated_timings(self) -> Vec<SpanTiming> {
        let mut entries: Vec<_> = self.accumulated_span_timings.into_iter()
            .collect();
        // By sorting the entries, we always obtain parents before child entries
        entries.sort();
        let mut timings = Vec::new();

        for (span_strings, duration) in entries {
            let timing = SpanTiming {
                span_name: span_strings.last().cloned().unwrap_or_else(|| String::new()),
                children: vec![],
                self_duration: duration
            };
            if span_strings.is_empty() {
                timings.push(timing);
            } else {
                insert_in_parent(timing, &span_strings[..span_strings.len() - 1], &mut timings);
            }
        }

        timings
    }
}

fn insert_in_parent(
    timing: SpanTiming,
    parent_strings: &[String],
    potential_ancestors: &mut Vec<SpanTiming>
) {
    if let Some(oldest_ancestor_name) = parent_strings.first() {
        if let Some(oldest_ancestor) = potential_ancestors.iter_mut().find(|span| span.name() == oldest_ancestor_name) {
            insert_in_parent(timing, &parent_strings[1..], &mut oldest_ancestor.children);
        } else {
            // There is no such ancestor, so let's create one
            // This shouldn't normally happen but could perhaps happen if for some reason
            // spans are mismatched in the log
            let mut ancestor = SpanTiming {
                span_name: oldest_ancestor_name.clone(),
                children: vec![],
                self_duration: StdDuration::ZERO,
            };
            insert_in_parent(timing, &parent_strings[1..], &mut ancestor.children);
        }
    } else {
        // the entry has no parent, so it must be a sibling of the potential ancestors
        potential_ancestors.push(timing)
    }
}