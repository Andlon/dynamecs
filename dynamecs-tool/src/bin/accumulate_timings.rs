use std::collections::HashMap;
use std::error::Error;
use std::{fmt, slice};
use std::io::{stdout, Write};
use std::time::Duration;
use tabwriter::TabWriter;
use dynamecs_analyze::iterate_records;
use dynamecs_analyze::timing::{accumulate_timings, SpanTiming2};

fn main() -> Result<(), Box<dyn Error>> {
    if let Some(arg) = std::env::args().skip(1).next() {
        let records: Vec<_> = iterate_records(&arg)?.collect::<Result<_, _>>()?;
        let timings = accumulate_timings(&records)?;

        // Map (parent, name) to durations, so that we can look up later
        let timings_map = compute_timings_map(&timings);
        let self_durations_map = compute_self_durations(&timings, &timings_map);

        let stdout = std::io::stdout().lock();
        let mut writer = TabWriter::new(stdout);
        writeln!(writer, "Duration\tSelf duration\tSelf %\t% of root\t% of parent\t Span")?;
        for timing in timings {
            let indent = {
                let pre_indent_num = timing.parent().len().saturating_sub(1);
                let mut indent = "│   ".repeat(pre_indent_num);
                if timing.parent().len() > 0 {
                    indent.push_str("├── ");
                }
                indent
            };
            let duration = format_duration(timing.duration());
            let root_ancestor = timing.parent().first();
            let root_duration = root_ancestor.and_then(|root| timings_map.get(slice::from_ref(root)).copied());
            let parent_duration = timings_map.get(timing.parent()).copied();
            let self_duration = self_durations_map.get(timing.span_path()).copied()
                .expect("Self duration must always be present");
            let self_duration_str = format_duration(self_duration);
            writeln!(writer, "{duration}\t{self_duration}\t{self_proportion}\t{root_proportion}\t{parent_proportion} \t {indent}{name}",
                     name = timing.name(),
                     self_duration = self_duration_str,
                     self_proportion = format_proportion(self_duration, Some(timing.duration())),
                     root_proportion = format_proportion(timing.duration(), root_duration),
                     parent_proportion = format_proportion(timing.duration(), parent_duration))?;
        }
        writer.flush()?;
    } else {
        eprintln!("No path to log file provided, exiting...");
    }
    Ok(())
}

fn compute_timings_map(timings: &[SpanTiming2]) -> HashMap<Vec<String>, Duration> {
    // Map (parent, name) to durations, so that we can look up later
    timings.iter()
        .map(|timing| {
            let mut path = timing.parent().to_vec();
            path.push(timing.name().to_string());
            (path, timing.duration())
        })
        .collect()
}

fn compute_self_durations(
    timings: &[SpanTiming2],
    timings_map: &HashMap<Vec<String>, Duration>
) -> HashMap<Vec<String>, Duration> {
    let mut self_durations_map = timings_map.clone();
    for timing in timings {
        if let Some(mut parent_self_duration) = self_durations_map.get_mut(timing.parent()) {
            *parent_self_duration -= timing.duration();
        }
    }
    self_durations_map
}

fn format_proportion(duration: Duration, other_duration: Option<Duration>) -> String {
    let proportion = other_duration.map(|relative_to| duration.as_secs_f64() / relative_to.as_secs_f64());
    proportion.map(|p| format!("{:5.1} %", p * 100.0))
        .unwrap_or_else(|| "N/A".to_string())
}

fn format_duration(duration: Duration) -> String {
    let s = duration.as_secs_f64();
    if s < 1e-7 {
        format!("{:7.2} ns", s * 1e9)
    } else if s < 1e-4 {
        format!("{:7.2} µs", s * 1e6)
    } else if s < 1e-1 {
        format!("{:7.2} ms", s * 1e3)
    } else {
        format!("{:7.2e}  s", s)
    }
}