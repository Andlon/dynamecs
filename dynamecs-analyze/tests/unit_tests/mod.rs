use serde_json::json;
use time::format_description::well_known::Iso8601;
use time::OffsetDateTime;
use dynamecs_analyze::{iterate_records_from_reader, Level, Record, RecordKind};
use dynamecs_analyze::timing::accumulate_timings;

/// Helper macro for succinctly creating a span path from a list of literals.
macro_rules! span_path {
    ($($strings:expr),*) => {
        SpanPath::new(vec![$($strings.to_string()),*])
    }
}

mod span_path;
mod span_tree;
mod timing;

#[test]
fn test_basic_records_iteration() {
    let log_data = r###"
        {"timestamp":"2023-03-29T12:48:50.213348Z","level":"TRACE","fields":{"message":"enter"},"target":"dynsys::backward_euler","span":{"name":"Backward Euler IP assemble"},"spans":[{"name":"run"},{"step_index":16,"name":"step"},{"name":"Backward Euler"},{"name":"Backward Euler"},{"hessian_mod":"NoModification","k":8,"name":"Newton iteration"},{"name":"Backward Euler IP assemble"}]}
        {"timestamp":"2023-03-29T12:48:51.440914Z","level":"INFO","fields":{"message":"exit"},"target":"dynsys::backward_euler","span":{"name":"hessian"},"spans":[{"name":"run"},{"step_index":16,"name":"step"},{"name":"Backward Euler"},{"name":"Backward Euler"},{"hessian_mod":"NoModification","k":8,"name":"Newton iteration"},{"name":"Backward Euler IP assemble"}]}
        {"timestamp":"2023-03-29T12:48:51.440972Z","level":"TRACE","fields":{"message":"exit"},"target":"dynsys::backward_euler","span":{"name":"Backward Euler IP assemble"},"spans":[{"name":"run"},{"step_index":16,"name":"step"},{"name":"Backward Euler"},{"name":"Backward Euler"},{"hessian_mod":"NoModification","k":8,"name":"Newton iteration"}]}
        {"timestamp":"2023-03-29T12:48:51.441519Z","level":"DEBUG","fields":{"message":"enter"},"target":"dynsys::backward_euler","span":{"name":"solve_linear_system"},"spans":[{"name":"run"},{"step_index":16,"name":"step"},{"name":"Backward Euler"},{"name":"Backward Euler"},{"hessian_mod":"NoModification","k":8,"name":"Newton iteration"},{"name":"solve_linear_system"}]}
    "###;
    let records: Vec<Record> = iterate_records_from_reader(log_data.as_bytes())
        .collect::<eyre::Result<_>>()
        .unwrap();

    assert_eq!(records.len(), 4);

    {
        let record = &records[0];
        assert_eq!(record.level(), Level::Trace);
        assert_eq!(record.target(), "dynsys::backward_euler");
        assert_eq!(record.kind(), RecordKind::SpanEnter);
        assert_eq!(record.message(), Some("enter"));
        assert_eq!(record.timestamp(), &OffsetDateTime::parse("2023-03-29T12:48:50.213348Z", &Iso8601::DEFAULT).unwrap());
        assert_eq!(record.span().unwrap().name(), "Backward Euler IP assemble");
        assert_eq!(record.span().unwrap().fields(), &json! {{
            "name": "Backward Euler IP assemble",
        }});
        assert_eq!(record.spans().unwrap().len(), 6);
        assert_eq!(record.spans().unwrap()[0].name(), "run");
        assert_eq!(record.spans().unwrap()[0].fields(), &json! {{
            "name": "run"
        }});
        assert_eq!(record.spans().unwrap()[1].name(), "step");
        assert_eq!(record.spans().unwrap()[1].fields(), &json! {{
            "name": "step",
            "step_index": 16
        }});
        assert_eq!(record.spans().unwrap()[2].name(), "Backward Euler");
        assert_eq!(record.spans().unwrap()[2].fields(), &json! {{
            "name": "Backward Euler"
        }});
        assert_eq!(record.spans().unwrap()[3].name(), "Backward Euler");
        assert_eq!(record.spans().unwrap()[3].fields(), &json! {{
            "name": "Backward Euler"
        }});
        assert_eq!(record.spans().unwrap()[4].name(), "Newton iteration");
        assert_eq!(record.spans().unwrap()[4].fields(), &json! {{
            "name": "Newton iteration",
            "hessian_mod": "NoModification",
            "k": 8,
        }});
        assert_eq!(record.spans().unwrap()[5].name(), "Backward Euler IP assemble");
        assert_eq!(record.spans().unwrap()[5].fields(), &json! {{
            "name": "Backward Euler IP assemble",
        }});
    }

    {
        let record = &records[1];
        assert_eq!(record.level(), Level::Info);
        assert_eq!(record.target(), "dynsys::backward_euler");
        assert_eq!(record.kind(), RecordKind::SpanExit);
        assert_eq!(record.message(), Some("exit"));
        assert_eq!(record.timestamp(), &OffsetDateTime::parse("2023-03-29T12:48:51.440914Z", &Iso8601::DEFAULT).unwrap());
        assert_eq!(record.span().unwrap().name(), "hessian");
        assert_eq!(record.span().unwrap().fields(), &json! {{
            "name": "hessian",
        }});
        assert_eq!(record.spans().unwrap().len(), 6);

        // TODO: Test the rest
    }
}

#[test]
fn test_basic_timing_collection() {
    let log_data = r###"
{"timestamp":"2023-03-29T18:16:50.073142Z","level":"INFO","fields":{"message":"enter"},"target":"dynamecs_app","span":{"name":"run"},"spans":[{"name":"run"}]}
{"timestamp":"2023-03-29T18:16:50.073564Z","level":"INFO","fields":{"message":"Starting simulation of scenario \"basic\""},"target":"dynamecs_app","span":{"name":"run"},"spans":[{"name":"run"}]}
{"timestamp":"2023-03-29T18:16:50.073929Z","level":"INFO","fields":{"message":"No time step configured. Using default dt = 0.016666666666666666"},"target":"dynamecs_app","span":{"name":"run"},"spans":[{"name":"run"}]}
{"timestamp":"2023-03-29T18:16:50.074164Z","level":"INFO","fields":{"message":"enter"},"target":"dynamecs_app","span":{"step_index":0,"name":"step"},"spans":[{"name":"run"},{"step_index":0,"name":"step"}]}
{"timestamp":"2023-03-29T18:16:50.074370Z","level":"DEBUG","fields":{"message":"Running post-systems for initial state"},"target":"dynamecs_app","span":{"step_index":0,"name":"step"},"spans":[{"name":"run"},{"step_index":0,"name":"step"}]}
{"timestamp":"2023-03-29T18:16:50.074565Z","level":"DEBUG","fields":{"message":"enter"},"target":"basic_app","span":{"something":"test","name":"system_that_logs"},"spans":[{"name":"run"},{"step_index":0,"name":"step"},{"something":"test","name":"system_that_logs"}]}
{"timestamp":"2023-03-29T18:16:50.074771Z","level":"DEBUG","fields":{"message":"some debug info","number":42},"target":"basic_app","span":{"something":"test","name":"system_that_logs"},"spans":[{"name":"run"},{"step_index":0,"name":"step"},{"something":"test","name":"system_that_logs"}]}
{"timestamp":"2023-03-29T18:16:50.074911Z","level":"DEBUG","fields":{"message":"exit"},"target":"basic_app","span":{"something":"test","name":"system_that_logs"},"spans":[{"name":"run"},{"step_index":0,"name":"step"}]}
{"timestamp":"2023-03-29T18:16:50.075098Z","level":"DEBUG","fields":{"message":"enter"},"target":"basic_app","span":{"something":"other test","name":"other_system_that_logs"},"spans":[{"name":"run"},{"step_index":0,"name":"step"},{"something":"other test","name":"other_system_that_logs"}]}
{"timestamp":"2023-03-29T18:16:50.075291Z","level":"DEBUG","fields":{"message":"some other debug info","digits":17},"target":"basic_app","span":{"something":"other test","name":"other_system_that_logs"},"spans":[{"name":"run"},{"step_index":0,"name":"step"},{"something":"other test","name":"other_system_that_logs"}]}
{"timestamp":"2023-03-29T18:16:50.075428Z","level":"DEBUG","fields":{"message":"exit"},"target":"basic_app","span":{"something":"other test","name":"other_system_that_logs"},"spans":[{"name":"run"},{"step_index":0,"name":"step"}]}
{"timestamp":"2023-03-29T18:16:50.075745Z","level":"INFO","fields":{"message":"Starting step 0 at simulation time 0.00000 (dt = 1.66667e-2)"},"target":"dynamecs_app","span":{"step_index":0,"name":"step"},"spans":[{"name":"run"},{"step_index":0,"name":"step"}]}
{"timestamp":"2023-03-29T18:16:50.075945Z","level":"DEBUG","fields":{"message":"enter"},"target":"basic_app","span":{"something":"test","name":"system_that_logs"},"spans":[{"name":"run"},{"step_index":0,"name":"step"},{"something":"test","name":"system_that_logs"}]}
{"timestamp":"2023-03-29T18:16:50.076217Z","level":"DEBUG","fields":{"message":"some debug info","number":42},"target":"basic_app","span":{"something":"test","name":"system_that_logs"},"spans":[{"name":"run"},{"step_index":0,"name":"step"},{"something":"test","name":"system_that_logs"}]}
{"timestamp":"2023-03-29T18:16:50.076413Z","level":"DEBUG","fields":{"message":"exit"},"target":"basic_app","span":{"something":"test","name":"system_that_logs"},"spans":[{"name":"run"},{"step_index":0,"name":"step"}]}
{"timestamp":"2023-03-29T18:16:50.076594Z","level":"DEBUG","fields":{"message":"enter"},"target":"basic_app","span":{"something":"other test","name":"other_system_that_logs"},"spans":[{"name":"run"},{"step_index":0,"name":"step"},{"something":"other test","name":"other_system_that_logs"}]}
{"timestamp":"2023-03-29T18:16:50.076780Z","level":"DEBUG","fields":{"message":"some other debug info","digits":17},"target":"basic_app","span":{"something":"other test","name":"other_system_that_logs"},"spans":[{"name":"run"},{"step_index":0,"name":"step"},{"something":"other test","name":"other_system_that_logs"}]}
{"timestamp":"2023-03-29T18:16:50.076916Z","level":"DEBUG","fields":{"message":"exit"},"target":"basic_app","span":{"something":"other test","name":"other_system_that_logs"},"spans":[{"name":"run"},{"step_index":0,"name":"step"}]}
{"timestamp":"2023-03-29T18:16:50.077044Z","level":"INFO","fields":{"message":"exit"},"target":"dynamecs_app","span":{"step_index":0,"name":"step"},"spans":[{"name":"run"}]}
{"timestamp":"2023-03-29T18:16:50.077210Z","level":"INFO","fields":{"message":"enter"},"target":"dynamecs_app","span":{"step_index":1,"name":"step"},"spans":[{"name":"run"},{"step_index":1,"name":"step"}]}
{"timestamp":"2023-03-29T18:16:50.077323Z","level":"INFO","fields":{"message":"exit"},"target":"dynamecs_app","span":{"step_index":1,"name":"step"},"spans":[{"name":"run"}]}
{"timestamp":"2023-03-29T18:16:50.077568Z","level":"INFO","fields":{"message":"Simulation ended"},"target":"dynamecs_app","span":{"name":"run"},"spans":[{"name":"run"}]}
{"timestamp":"2023-03-29T18:16:50.077681Z","level":"INFO","fields":{"message":"exit"},"target":"dynamecs_app","span":{"name":"run"},"spans":[]}
    "###;
    let records: Vec<_> = iterate_records_from_reader(log_data.as_bytes())
        .collect::<eyre::Result<_>>()
        .unwrap();
    let timings = accumulate_timings(&records).unwrap();
    insta::assert_debug_snapshot!(timings);
}