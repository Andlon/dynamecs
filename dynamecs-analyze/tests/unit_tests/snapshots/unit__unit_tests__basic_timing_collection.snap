---
source: dynamecs-analyze/tests/unit_tests/mod.rs
expression: timings
---
[
    SpanTiming {
        span_name: "run",
        children: [
            SpanTiming {
                span_name: "step",
                children: [
                    SpanTiming {
                        span_name: "other_system_that_logs",
                        children: [],
                        self_duration: 652µs,
                    },
                    SpanTiming {
                        span_name: "system_that_logs",
                        children: [],
                        self_duration: 814µs,
                    },
                ],
                self_duration: 2.993ms,
            },
        ],
        self_duration: 4.539ms,
    },
]
